### ISSUE TITLE
AuthHttp stripping error response codes on catch

### ISSUE THREAD

**Issue by paynoattn** [auth0/angular2-jwt#227]
Currently in my project given I have a dataservice doing the following:

``` 
export class DataService {
  constructor( private authSvc: AuthHttp ) {}

  authCheck() {
    return tokenNotExpired();
  }

    get(url: string) {
    let headers = this.makeHeaders();
    let options = new RequestOptions({ headers: headers });
    if (this.authCheck()) {
      return this.authSvc.get(url, options)
                         .catch(this.handleError)
                         .map(this.extractData)
    } else {
      this.handleAuthFailure();
    }
  }
}
```

I have noticed that in this case the handleError always recieves a response where response.status is 0.

This is not the case when using the default Angular Http service and inserting the auth headers manually.

**Comment by escardin** [comment]
AuthHttp does not transform the response. It sounds to me like you have a malformed request somehow. 

Can you compare the requests and responses in the network inspector of your browser and let us know of any differences?

**Comment by paynoattn** [comment]
Actually this is happening in the angular2 Http service as well, but only for 404 responses. This isn't an authhttp problem.

**Comment by escardin** [comment]
@paynoattn are you sure they're 404's and not network errors of some sort? There's an issue for http about that, I don't know if it got resolved.

**Comment by paynoattn** [comment]
@escardin actually this is a deeper issue I'm going to have to research. When using Authhttp or http with auth headers sent from a ```localstorage.get()``` this happens for any error. I am sure that these are errors sent back from my API as I see the error codes being sent by my server back in the logs, and them matching in my network tab. 

Here's my full data service:

```
import { Injectable } from '@angular/core';
import { Headers, RequestOptions, Response } from '@angular/http';
import { Observable } from 'rxjs/Observable';
import { AuthHttp, tokenNotExpired } from 'angular2-jwt';
import { Store } from '@ngrx/store';
import { Router } from '@angular/router';

import { FlashMessage } from '../models';

@Injectable()

export class DataService {
  private errorMsg: string = 'Server error, please try reloading your browser!';

  constructor(
    private authSvc: AuthHttp,
    private _router: Router,
    private _store: Store<any>
  ) { }

  authCheck() {
    return tokenNotExpired();
  }

  handleAuthFailure() {
      this._router.navigate(['/']);
      let myErr: FlashMessage = {
        type: 'danger',
        text: 'You are not logged in!'
      };
      this._store.dispatch({type: 'ADD_MESSAGE', payload: myErr});
  }

  makeHeaders() {
    let headers = new Headers();
    headers.append( 'Content-Type', 'application/json' );
    headers.append( 'Accept', 'application/json' );
    return headers;
  }

  get(url: string) {
    let headers = this.makeHeaders();
    let options = new RequestOptions({ headers: headers });
    return this.authSvc.get(url, options)
      .map(this.extractData)
      .catch(this.handleError);
  }

  ...

  private extractData(res: Response) {
    // console.log(res);
    let body = res.json();
    return body || { };
  }

  private handleError(error: Response | any) {
        console.log('This is error: ', error.status);
        let errMsg = (error.message) ? error.message :
        error.status ? `${error.status} - ${error.statusText}` : this.errorMsg;
        console.error(error, errMsg); // log to console instead
        return Observable.throw(errMsg);
    }
}

```

Note that this also happens when I switch to Http and send the Auth headers like so.

```
...

makeHeaders() {
    let headers = new Headers();
    headers.append( 'Content-Type', 'application/json' );
    headers.append( 'Accept', 'application/json' );
    headers.append( 'Authorization', localstorage.getItem('id-token');
    return headers;
  }

  get(url: string) {
    let headers = this.makeHeaders();
    let options = new RequestOptions({ headers: headers });
    if (this.authCheck()) {
      return this.http.get(url, options)
        .map(this.extractData)
        .catch(this.handleError);
    } else {
      this.handleAuthFailure();
    }
  }

...
```

But if I simply do not send the auth headers and disable the auth_check in my backend, the response headers are correctly preserved.

And my error stack.

![screen shot 2016-12-06 at 12 17 13 pm](https://cloud.githubusercontent.com/assets/7308577/20940661/d7ef71ae-bbb0-11e6-9641-cfd457502264.png)

**Comment by escardin** [comment]
AHA! A CORS error! It becomes clear!

So, the reason you don't have a response code is the network error stuff from before. The reason you have a 'network' error is because for XHR requests need a preflight OPTIONS request to be sent so that the browser is sure you can actually make that request (the options is sent by the browser, nothing required on the browser end on your part).

You will need to google CORS (general info on CORS), and CORS for your platform (to solve the specific issue).

It boils down to allowing OPTIONS and responding with the header noted missing in your request.

You're probably getting 404's because your server isn't allowing OPTIONS methods or something. It might be that the method is blocked when you don't have auth. I believe that options won't send your authorization header.

It's been a little while since I've debugged such an issue, but you should see an OPTIONS request in the browser network tab (Only once per api endpoint or some such) and in your server logs

**Comment by paynoattn** [comment]
Good catch, I can't believe I ddin't notice that. It appears for whatever reason my backend is sending cors headers on 401s but not 404s / 500s. Hopefully this will resolve it.

**Comment by brassier** [comment]
I'm running into the same thing here.  Our back-end is intentionally returning a 404 in some cases, and I also get a status 0 when using AuthHttp.  Using Angular's HTTP component does result in valid HTTP status codes.

Are you sure this isn't an issue?  I'm going to open a new issue since this one is closed.
